src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/PoolingDataSource.cs

```diff
@@ -425,12 +438,55 @@ internal void PruneIdleConnections()
         }
 
         /// <summary>
-        /// Warms up the pool to bring it up to min pool size.
+        /// Warms up the pool by bringing it up to min pool size.
         /// </summary>
-        /// <exception cref="NotImplementedException"></exception>
-        internal void WarmUp()
+        /// <returns>A ValueTask representing the warmup process.</returns>
+        internal ValueTask WarmUp()
         {
-            throw new NotImplementedException();
+            lock (_warmupLock)
```

Comment:
Locks can be cheap until they are not. 
Consider moving away from the usage of locks in a method which is meant to be run asynchronously.